internal/controller/stages/verification.go

```diff
@@ -0,0 +1,433 @@
+// are not available on the shards.
+func (r *reconciler) getAnalysisRunNamespace(stage *kargoapi.Stage) string {
+	if r.shardName == "" {
+		return stage.Namespace
+	}
+	return "analysis-runs" // TODO: KR: Do not hardcode this
+}
+
+func (r *reconciler) buildAnalysisRun(
+	stage *kargoapi.Stage,
+	templates []*rollouts.AnalysisTemplate,
+) (*rollouts.AnalysisRun, error) {
+	// maximum length of the stage name used in the promotion name prefix before it exceeds
+	// kubernetes resource name limit of 253
+	// 253 - 1 (.) - 26 (ulid) - 1 (.) - 7 (sha) = 218
+	const maxStageNamePrefixLength = 218
+
+	// Build the name of the AnalysisRun
+	shortHash := stage.Status.CurrentFreight.ID
+	if len(shortHash) > 7 {
+		shortHash = shortHash[0:7]
+	}
+	shortStageName := stage.Name
+	if len(stage.Name) > maxStageNamePrefixLength {
+		shortStageName = shortStageName[0:maxStageNamePrefixLength]
+	}
+	analysisRunName := strings.ToLower(fmt.Sprintf("%s.%s.%s", shortStageName, ulid.Make(), shortHash))
+
+	// Build the labels and annotations for the AnalysisRun
+	var numLabels int
+	var numAnnotations int
+	if stage.Spec.Verification.AnalysisRunMetadata != nil {
+		numLabels = len(stage.Spec.Verification.AnalysisRunMetadata.Labels)
+		numAnnotations = len(stage.Spec.Verification.AnalysisRunMetadata.Annotations)
+	}
+	// Kargo will add up two lbls of its own, so size the map accordingly
+	lbls := make(map[string]string, numLabels+2)
+	annotations := make(map[string]string, numAnnotations)
+	if stage.Spec.Verification.AnalysisRunMetadata != nil {
+		for k, v := range stage.Spec.Verification.AnalysisRunMetadata.Labels {
+			lbls[k] = v
+		}
+		for k, v := range stage.Spec.Verification.AnalysisRunMetadata.Annotations {
+			annotations[k] = v
+		}
+	}
+	lbls[kargoapi.StageLabelKey] = stage.Name
+	lbls[kargoapi.FreightLabelKey] = stage.Status.CurrentFreight.ID
+
+	// Flatten templates into a single template
+	template, err := flattenTemplates(templates)
+	if err != nil {
+		return nil, errors.Wrap(err, "error flattening templates")
+	}
+
+	// Merge the args from the template with the args from the Stage
+	rolloutsArgs := make([]rollouts.Argument, len(stage.Spec.Verification.Args))
+	for i, argument := range stage.Spec.Verification.Args {
+		arg := argument // Avoid implicit memory aliasing
+		rolloutsArgs[i] = rollouts.Argument{
+			Name:  arg.Name,
+			Value: &arg.Value,
+		}
+	}
+	mergedArgs, err := mergeArgs(template.Spec.Args, rolloutsArgs)
+	if err != nil {
+		return nil, errors.Errorf("error merging arguments")
+	}
+
+	return &rollouts.AnalysisRun{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        analysisRunName,
+			Namespace:   r.getAnalysisRunNamespace(stage),
+			Labels:      lbls,
+			Annotations: annotations,
+		},
+		Spec: rollouts.AnalysisRunSpec{
+			Metrics:              template.Spec.Metrics,
+			DryRun:               template.Spec.DryRun,
+			MeasurementRetention: template.Spec.MeasurementRetention,
+			Args:                 mergedArgs,
```

Comment:
Because we don't want to tie Kargo with certain rollout versions, and we don't want to keep making sure our rollout types are up to date, one (future) improvement we should do is to copy the analysis template as unstructured and just have code that merges the args (which we would still strongly type).

Alternatively, we define part of our rollout type as RawMessage so that we don't care about anything other than args.

But this can be a future improvement.